### ISSUE TITLE
Tasks introduced via 'sync' should trigger hooks

### ISSUE THREAD

**Issue by johl** [GothenburgBitFactory/taskwarrior#2142]
When using several instances of Taskwarrior on different devices (e.g. on a mobile phone with Android and on a computer), only locally added tasks trigger the on-modify hook. 

Syncing means that tasks can be added from other sources and on-modify should be triggered when a new task comes in from a remote source. I see the exception made for sync to be counter-intuitive and cannot see a good reason for it.

**Comment by pbeckingham** [comment]
Sorry, disagree.  Tasks arriving via sync have already passed through any on-modify hook on the CLI, and shouldn't be modified again.

**Comment by johl** [comment]
This would be true if the only possibility to add tasks would be to add them with the CLI or on a system that has hooks. This is not the case.

**Comment by sherwoodinc** [comment]
I was about to report this same issue.

I just started using taskwarrior+timewarrior (with the onModify hook to track time) yesterday and set up sync, also installing the android app.
If I start/stop tasks in the android app, and sync both devices, my desktop computer hooks don't get executed, so timewarrior doesn't update time tracking.
This is unfortunate and I don't see any workaround.

**Comment by tbabej** [comment]
This is a pickle, because while this is a valid use case (i.e. a mobile TW client might not have a way of implementing hooks), it also means that the hooks scripts now can run multiple times for any given task (i.e. if the hooks are installed on two clients).

In that case we'd either have to require the hook implementation to be idempotent or implement some mechanism for that ourselves.

**Comment by johl** [comment]
Any news on this?

**Comment by tbabej** [comment]
As things stand, we don't have a good solution for this problem. Maybe a new hook entrypoint, `on-sync` could help here?

**Comment by johl** [comment]
That would absolutely help. I think that's the best solution.

**Comment by sherwoodinc** [comment]
I "fixed" the issue in my desktop install by forcing on-modify for synced tasks.
In my simple workflow it doesn't bring issues.

About the on-sync hook (which sounds good)... Does it mean timewarrior should be hooked both to on-modify and on-sync to work properly in my mobile+desktop setup?

**Comment by tbabej** [comment]
Correct. You could just create a symlink to the `on-modify` hook. By doing that however, you should make sure the underlying hook is idempotent (meaning it's fine if it runs on the same task twice), as you can run into the problem where it would be first executed on first client during `on-modify` event and then again on the second client during `on-sync` event.

**Comment by tbabej** [comment]
Related to #1790.

**Comment by djmitche** [comment]
I think sync and hooks are fundamentally incompatible, since the hooks happen on one replica, and sync is between many replicas.

In practice, I think this is OK -- most people use one or the other. But I can see the opportunity for lots of trouble in an `on-sync` hook, as if that hook creates any further modifications then those, too, would get sync'd to other replicas, easily causing loops.

**Comment by goodevilgenius** [comment]
> I "fixed" the issue in my desktop install by forcing on-modify for synced tasks.

@sherwoodinc I know you commented on this issue several years ago, but I was wondering if you could explain how you did this. I'm looking to make this work for my use case. I'd really like to get synced tasks to trigger my tasks.

**Comment by goodevilgenius** [comment]
I'd like to suggest one of two options:

1. Add two new hooks: `on-syncadd` and `on-syncmodify` that would mirror the `on-add` and `on-modify` hooks exactly. They would run once for every task added or modified during a single sync. Importantly, they wouldn't be able to modify the task itself (like `on-exit`), since that should have been done on the originating replica.
2. A configuration option that would cause `on-add` and `on-modify` to run for synced tasks, which would default to off.

I think either of these options would work for those of us who need this feature.
